remove arbitrary 600 pixel limit on quick input widget

dscherger opened the pull request:
Add maximumWidth and relativeWidth settings to customize width of quickInput widget rather than the hardcoded minimum of 62% of available width or 600 pixel max width

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #85374

dscherger committed:
remove arbitrary 600 pixel limit on quick input widget

use 75% of available width rather than minimum of 62% of
available width or 600 pixel max width

chrmarti commented:
@dscherger Wouldn't this make it super-wide on wide screens? I think there is value in having it "in the middle" for easier reading and visual appeal.

dscherger commented:
Hi @chrmarti thanks for looking at this! 

It does make the quick input widget 75% of the width of the vscode window, still "in the middle" of the window but wider than it is now.  Perhaps 75% is too much but for projects with the unfortunate combination of long (40-50 chars), non-unique file names and with paths that have the distinguishing element somewhere beyond the 30 character mark there's currently no way to see which variant of a file to open.

I'm sure there other are changes that could be made to solve this problem but it has been reported several times as #38225, #85374, #90948, #100214 and it would be nice to have a workaround in the meantime.

Other Solutions:

* display the full path of the currently selected item somewhere else (tool tip, bottom of window, etc.) without length constraints
* move the ... from the end of the displayed path to the front, so that trailing distinguishing elements can be displayed (may fail depending on where the unique element is)
* remove display of the filename from the front of the path and display the full path (may still fail for long paths)
* allow left/right scrolling of the content within the quick picker to help with navigation & selection (probably the best solution for very long paths)
* allow the quick picker to resize dynamically to fit the full path of the currently available elements (may still fail for very long paths)

Improvements to this Solution:

We could preserve the current look by default and add settings to control both the max width (default to 600px but allow unlimited) and the scale factor (defaulted to 62% but allow for 100%). This would allow people who need it to deal with the problem and not affect anyone who doesn't. Of course this may still fail for really long paths but it will handle less extreme situations a little better than they are now.

dscherger committed:
add maximumWidth and relativeWidth quickOpen settings

These preserve the existing behaviour and width of the quickOpen widget but
allow for customization in situations where long filenames and paths can cause
problems selecting the correct file to open.

TylerLeonhardt commented:
Thank you so much for this!

Although the settings are a solution, I don't think they're the best solution for this task. I'd prefer to see the quick input dynamically expand with the size of the content or allow it to be resized.

I think we'll want to revisit this when we look at the new "omnibar" experience detailed in https://github.com/microsoft/vscode/issues/115641

an-dr-eas-k commented:
This is a fairly minor change that replaces the customize-ui workaround for many users (with all its security implications). The merge broadens the user-base that will give feedback, maybe also usable for the future omnibar.

PalGenadich commented:
@isidorn Please push this through. This was my case for using Customize UI. And since 1.74 it's no longer viable https://github.com/iocave/customize-ui/issues/156

Moghul commented:
@chrmarti Aesthetics are important but not to the degree that they get in the way of functionality. If I can't use the tool well, it doesn't matter to me how pretty it is.

@dscherger Thank you very much for the PR, I look forward to it being merged! The only thing I mildly disagree with is the equally arbitrary 2000 maximum limit. Just let me set it to what I want. If it breaks because the number's too big, the settings can always be changed back.

dscherger commented:
> @chrmarti Aesthetics are important but not to the degree that they get in the way of functionality. If I can't use the tool well, it doesn't matter to me how pretty it is.
> 
> @dscherger Thank you very much for the PR, I look forward to it being merged! The only thing I mildly disagree with is the equally arbitrary 2000 maximum limit. Just let me set it to what I want. If it breaks because the number's too big, the settings can always be changed back.

@chrmarti @Moghul 
This is definitely a "perfect is the enemy of the good" type of situation.  It's been over 2 years since I put up the original fix and I'd be really curious to know how many people would have benefitted from it over that time while we're waiting for  omnibar perfection.  I haven't read the latest comments in https://github.com/microsoft/vscode/issues/115641 and don't have any sense of whether this is close to landing or if it will ever land.

As far as the arbitrary 2000 character limit goes, I'm not sure if this is a required value in the config or whether it can be left out entirely. It looks like I need to go resolve some conflicts and I'll have a look at the limit while rebasing.

nschikora commented:
Adding a screenshot here as well to make clear what this is about. The paths in the project I’m working on are not too deep but still line 3 and 4 are indistinguishable from each other. (One is an interface and the other the implementation). I run into this few times a day.

<img width="605" alt="image" src="https://github.com/microsoft/vscode/assets/9056560/51a2df19-0b30-4a14-8ca1-9fe9760a244c">

johnnyoshika commented:
Here's an example why this enhancement is desperately needed

![image](https://github.com/microsoft/vscode/assets/504505/81ffb90d-7b6c-4294-9bfb-b64a6810fae4)

Moghul commented:
What exactly has to happen for this to be merged and released? It seemed like such a slam dunk, but it's been 3 years.

protoEvangelion commented:
@aiday-mar @jrieken @amunger @meganrogge @connor4312 What needs to be done to get this merged? Pinging you because you are the top 5 contributors in the last month, we haven't heard back from maintainers in over 3 years, & there is overwhelming community desire for this. FYI, this small PR would solve my #1 gripe with VSCode: _I cannot quick search my files because I cannot see the full path 99% of the time._

johnnyoshika commented:
I would desperately love this as well 🙏

AaronBeaudoin commented:
The quick input is really useless in my repo for finding files without this change since everything is so deeply nested and I can see the paths. VA Code core team—please merge this!!

dscherger commented:
I've been bumping into this myself a lot recently and it's really made me sad that this has never been merged. I've updated the branch, fixed some conflicts and re-tested and it still seems to work fine. This has been waiting now for almost 4 years and it would be great to see it merged even as a temporary solution. I'd be more than happy to see this deleted once a better approach is available but in the meantime, please can we get this merged?

@Tyriar @meganrogge @aiday-mar @jrieken @TylerLeonhardt what do we need to do to get this approved?

johnnyoshika commented:
This is desperately needed. When working in a large monorepo with deeply nested files, the 600px limit is very restrictive.

vecuenca commented:
This would be a huge QOL improvement! 😀

Tyriar reviewed (changes requested):
@TylerLeonhardt I've tested this and it works well, I think we should move forward and merge it instead of waiting for the ideal solution. If we do end up getting dynamic sizing happening we can always remove these settings at that point.

Tyriar committed:
Tweak min/max values

Tyriar committed:
Make experimental and clamp numbers

Tyriar commented on src/vs/workbench/browser/workbench.contribution.ts:
```diff
@@ -490,6 +490,20 @@ const registry = Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Con
+				'description': localize('workbench.quickOpen.maxWidth', "Limits the maximum width of Quick Open to the specified number of pixels."),
+				'default': 600,
+				'minimum': 200,
+				'maximum': 2000
```
Leaving it for now

protoEvangelion reviewed (approved):
Tested and works as expected:

### BEFORE

![image](https://github.com/user-attachments/assets/b1374ab2-d22a-41fb-8e62-4dd7173f6166)

### AFTER

![image](https://github.com/user-attachments/assets/058ac062-8908-4ec2-932e-309ed7f122db)

protoEvangelion commented:
@Tyriar are we ready to rock on this?

Tyriar commented:
@protoEvangelion spoke with some people in the team and we're a little concerned about adding a setting that will just get deleted. I think the way forward here is to allow the user to resize this by using a `Sash` on the right side of the quick pick.

You can see a similar interaction in "quick chat" if you have copilot installed:

![Recording 2024-11-11 at 10 08 55](https://github.com/user-attachments/assets/9eaf09d0-695e-45bb-8a30-4b87cf3d21d5)

We would then just want it to be horizontal. How this gets persisted is a bit of an unknown to me still, but my guess is maybe we want to store it in the percentage-based value and do away with the pixel one all together? So most users will configure it purely through the mouse interaction.

johnnyoshika commented:
I desperately need this. If we can push out a temporary fix, it would be such a QOL improvement. I've been dealing with this pain in my large monorepo for over 3 years. Having come from Visual Studio IDE (which allows resizing of the command palette), this has been a tough adjustment.

wadamczyk-insolutions commented:
Is there any way we can help push this forward? This issue is currently very frustrating for our projects. From what I understand, the only remaining step is obtaining Community PR Approvals. Could you explain the process for this? Is there any way to expedite these approvals?

vivodi commented:
Is this still relevant?

Moghul commented:
Is there any reason this wouldn't be relevant? Did a feature come out to make this irrelevant? Have all monorepos been deprecated?

Q-back commented:
It's 2025 now, let's sum up where we are.
This fix was introduced ~4 years ago, the issue is still there.
This solution was proven to fix the issue.
It was **rejected** because there are potentially better solutions (e.g. the [omnibar](https://github.com/microsoft/vscode/issues/115641) that was never merged as well since 2021).
Also, maintainers are worried about introducing changes that would be later deleted.
Although these changes would stay here 4 years if they'd merged back in 2021.

So with that in mind maybe it would be best to merge it despite everything because other solutions did not work anyway.

These settings would help many people.
If these settings will be `experimental` you don't need to sorry anyone for deleting them later when something better drops.

utilityboy commented:
There is now the ability to move the quick picker around as of `1.97.0`.  I would be curious to see the metrics to understand the importance of this enhancement vs. the option to adjust the width.

![image](https://github.com/user-attachments/assets/7f7f6ce8-ba28-4eaa-989e-f1f5e3c62860)

Moghul commented:
Personally, I could not care less where the menu is, I guess I wouldn't have it on the right.
But this feature... To me, it's mandatory, and it's not there. It's been long enough that I might start looking at another IDE.

1803200 reviewed (commented):
-

olhapi commented:
There's clearly a prioritization issue. 600px wide `quickInput` looks especially great on a 4k or ultra wide displays.

wadamczyk-insolutions commented:
Come on. There are just 24 changed lines waiting to be reviewed — it can't really be that hard, can it? It's mildly infuriating to have to edit the CSS files by hand after every Visual Studio Code update just to fix this. Meanwhile, the ability to change the input widget's position has somehow made it into a release. Does anyone know who can be pinged to finally close this issue?

MasterInQuestion commented:
    In fact it should just be:
    `min-width: min( 600px, 62% ); max-width: 90%;`
    .
    But seems the "flex" layout in its child elements glitched somehow:
    Not properly reflects the filenames width.
    (flexbox layout in CSS seems to be generally broken; use "&lt;table>" instead?)

wadamczyk-insolutions commented:
Well, come on! What can be done to get this finally fixed after so many years? It's a simple task! I'm happy to help... it's so frustrating to still need to change the CSS manually after every update and have VS Code complain it's corrupted after every restart...
